kedro/contrib/io/pyspark/spark_hive_data_set.py

```diff
@@ -0,0 +1,305 @@
+
+        Raises:
+            DataSetError: Invalid configuration supplied
+        """
+        self._database = database
+        self._table = table
+        self._stage_table = "_temp_" + table
+        self._valid_write_modes = ["insert", "upsert", "overwrite"]
+        if write_mode not in self._valid_write_modes:
+            raise DataSetError(
+                "Invalid write_mode provided: {invalid}. Write_mode must be one of {valid}".format(
+                    invalid=write_mode, valid=self._valid_write_modes
+                )
+            )
+        self._write_mode = write_mode
+        if self._write_mode == "upsert" and not table_pk:
+            raise DataSetError("table_pk must be set to utilise upsert read mode")
+        self._table_pk = table_pk
+
+        self._table_columns = self._load().columns if self._exists() else None
+
+        if (
+            self._table_pk
+            and self._exists()
+            and set(self._table_pk) - set(self._table_columns)
+                (col_name in self._table_columns for col_name in self._table_pk)
+            )
+        ):
+            raise DataSetError(
+                "columns [{colnames}] selected as PK not found in table {database}.{table}".format(
+                    colnames=", ".join(sorted(set(self._table_pk) - set(self._table_columns))),
+                    database=self._database,
+                    table=self._table,
+                )
+            )
+
+    @staticmethod
+    def _get_spark():
+        return SparkSession.builder.getOrCreate()
+
+    def _create_empty_hive_table(self, data):
+        data.createOrReplaceTempView("tmp")
+        self._get_spark().sql(
+            "create table {database}.{table} select * from tmp limit 1".format(
+                table=self._table, database=self._database
+            )
+        )
+        self._get_spark().sql(
+            "truncate table {database}.{table}".format(
+                database=self._database, table=self._table
+            )
+        )
+
+    def _load(self) -> DataFrame:
+        if not self._exists():
+            raise DataSetError(
+                "requested table not found: {database}.{table}".format(
+                    database=self._database, table=self._table
+                )
+            )
+        return self._get_spark().sql(
+            "select * from {database}.{table}".format(
+                database=self._database, table=self._table
+            )
+        )
+
+    def _save(self, data: DataFrame) -> None:
+        if not self._exists():
+            self._create_empty_hive_table(data)
+            self._table_columns = data.columns
+        self._validate_save(data)
+        write_methods = {
+            "insert": self._insert_save,
+            "upsert": self._upsert_save,
+            "overwrite": self._overwrite_save,
+        }
+        write_methods[self._write_mode](data)
+
+    def _insert_save(self, data):
```

Comment:
For consistency, can you please add type annotations for `data` arguments where missing?

Reply:
done in 1f672a0